go-controller/pkg/ovn/gateway.go

```diff
@@ -760,3 +763,167 @@ func (gw *GatewayManager) addExternalSwitch(prefix, interfaceID, nodeName, gatew
+		return fmt.Errorf("failed to delete logical switch port %s from switch %s: %w", portName, sw.Name, err)
+	}
+
+	// Remove the logical router port on the gateway router that connects to the join switch
+	logicalRouter := nbdb.LogicalRouter{Name: gw.gwRouterName}
+	logicalRouterPort := nbdb.LogicalRouterPort{
+		Name: types.GWRouterToJoinSwitchPrefix + gw.gwRouterName,
+	}
+	err = libovsdbops.DeleteLogicalRouterPorts(gw.nbClient, &logicalRouter, &logicalRouterPort)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete port %s on router %s: %w", logicalRouterPort.Name, gw.gwRouterName, err)
+	}
+
+	// Remove the static mac bindings of the gateway router
+	err = gateway.DeleteDummyGWMacBindings(gw.nbClient, gw.gwRouterName)
+	if err != nil {
+		return fmt.Errorf("failed to delete GR dummy mac bindings for node %s: %w", gw.nodeName, err)
+	}
+
+	// Remove the gateway router associated with nodeName
+	err = libovsdbops.DeleteLogicalRouter(gw.nbClient, &logicalRouter)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete gateway router %s: %w", gw.gwRouterName, err)
+	}
+
+	// Remove external switch
+	err = libovsdbops.DeleteLogicalSwitch(gw.nbClient, gw.extSwitchName)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete external switch %s: %w", gw.extSwitchName, err)
+	}
+
+	exGWexternalSwitch := types.EgressGWSwitchPrefix + gw.extSwitchName
+	err = libovsdbops.DeleteLogicalSwitch(gw.nbClient, exGWexternalSwitch)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		return fmt.Errorf("failed to delete external switch %s: %w", exGWexternalSwitch, err)
+	}
+
+	// This will cleanup the NodeSubnetPolicy in local and shared gateway modes. It will be a no-op for any other mode.
+	gw.delPbrAndNatRules(gw.nodeName, nil)
+	return nil
+}
+
+func (gw *GatewayManager) delPbrAndNatRules(nodeName string, lrpTypes []string) {
+	// delete the dnat_and_snat entry that we added for the management port IP
+	// Note: we don't need to delete any MAC bindings that are dynamically learned from OVN SB DB
+	// because there will be none since this NAT is only for outbound traffic and not for inbound
+	mgmtPortName := util.GetK8sMgmtIntfName(gw.netInfo.GetNetworkScopedName(nodeName))
+	nat := libovsdbops.BuildDNATAndSNAT(nil, nil, mgmtPortName, "", nil)
+	logicalRouter := nbdb.LogicalRouter{
+		Name: gw.clusterRouterName,
+	}
+	err := libovsdbops.DeleteNATs(gw.nbClient, &logicalRouter, nat)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
+		klog.Errorf("Failed to delete the dnat_and_snat associated with the management port %s: %v", mgmtPortName, err)
+	}
+
+	// delete all logical router policies on ovn_cluster_router
+	gw.removeLRPolicies(nodeName, lrpTypes)
+}
+
+func (gw *GatewayManager) staticRouteCleanup(nextHops []net.IP) {
+	ips := sets.Set[string]{}
```

Comment:
short circuit and return if theres no nextHops ? It can happen in L2 topology and its pointless iterating through all the LRSRs. Add comment about this.